optima/geospatial.py

```diff
@@ -0,0 +1,335 @@
+from Queue import Queue
+from PyQt4.QtCore import *
+from PyQt4.QtGui import *
+
+def geogui():
+    '''
+    Open the GUI for doing geospatial analysis.
+    
+    Version: 2016jan23
+    '''
+
+    global geoguiwindow, portfolio, projectslist, objectives, objectiveinputs, oldstdout
+    oldstdout = sys.stdout      # Make sure old stdout is remembered for redirection purposes.    
+    
+    portfolio = None
+    projectslist = []
+    objectives = defaultobjectives()
+    
+    ## Set parameters
+    wid = 650.0*2
+    hei = 550.0
+    top = 20
+    spacing = 40
+    left = 20.
+    projext = '.prj'
+    portext = '.prt'
+    
+    # Extending the normal widget class so that stdout can be restored upon closure. Pretty funky, no?
+    class GAWidget(QtGui.QWidget):
+        def __init__(self):
+            QtGui.QWidget.__init__(self)
+    
+        def closeEvent(self, event):
+            reply = QtGui.QMessageBox.question(self, 'Message',
+                "Are you sure to quit?", QtGui.QMessageBox.Yes, QtGui.QMessageBox.No)
+    
+            if reply == QtGui.QMessageBox.Yes:
+                sys.stdout = oldstdout  # The important bit.
+                event.accept()
+            else:
+                event.ignore()    
+    
+    ## Housekeeping
+    fig = figure(); close(fig) # Open and close figure...dumb, no? Otherwise get "QWidget: Must construct a QApplication before a QPaintDevice"
+    geoguiwindow = GAWidget() # Create panel widget
+    geoguiwindow.setGeometry(100, 100, wid, hei)
+    geoguiwindow.setWindowTitle('Optima geospatial analysis')
+    projectslist = []
+    
+    ##############################################################################################################################
+    ## Define functions
+    ##############################################################################################################################
+
+    
+    def _loadproj():
+        ''' Little helper function to load a project, since used more than once '''
+        filepath = QtGui.QFileDialog.getOpenFileNames(caption='Choose project file', filter='*'+projext)
+        project = None
+        try: 
+            project = loadobj(filepath, verbose=0)
+        except: 
+            print('Could not load file "%s"' % filepath)
+            return None
+        try: 
+            assert type(project)==Project
+            return project
+        except: 
+            print('File "%s" is not an Optima project file' % filepath)
+            return None
+        
+        
+    def makesheet():
+        ''' Create a geospatial spreadsheet template based on a project file '''
+        
+        ## 1. Load a project file
+        project = _loadproj()
+            
+        ## 2. Get destination filename
+        spreadsheetpath = QtGui.QFileDialog.getSaveFileName(caption='Save geospatial spreadsheet file', filter='*.xlsx')
+        
+        ## 3. Extract data needed from project (population names, program names...)
```

Comment:
is this not ready yet? if so, please add a todo in the code so it's clear whether there are pieces missing here